### DISCUSSION TITLE
How to get referer on backend in context

### DISCUSSION THREAD

**Discussion by paymog** [enisdenjo/graphql-ws#358] [Q&A] [answered]
I'm creating a websocket client using NextJS with the following:

```
    createClient({
        url: settings?.graphql?.wsUri,
        retryAttempts: 5,
        // more or less copied directly from https://github.com/enisdenjo/graphql-ws#retry-strategy
        retryWait: async () => {
          await waitForHealthCheck();

          // after the server becomes ready, wait for a second + random 100 - 200 ms timeout
          // (avoid DDoSing) and try connecting again, but only in production
          const delay = isProduction ? 100 + Math.random() * 100 : 10;
          await new Promise((resolve) => setTimeout(resolve, delay));
        },
     })
```

and on the backend I'm making a websocket server like so (some code omitted for general clarity):
```
useServer(
    {
      schema,
      context: async (ctx, msg, args) => {
        const context = {};

        if (ctx.extra.request.headers.referer) {
          console.log('found referer for websocket', ctx.extra.request.headers.referer);
        }

        const cookies = ctx.extra.request.headers.cookie?.split(';') || [];

        for (let i = 0; i < cookies.length; i++) {
          const cookie = cookies[i];
          const [key, value] = cookie.trim().split('=');
          if (key === JWT_COOKIE_KEY) {
            context.user = await getUserWithToken({ token: value });
            break;
          }
        }

        return context;
      },
    },
    wsServer,
  );
```

I'm finding that the backend never logs `found referer for websocket` and when I inspect the socket creation in the browser I see that no referer is being sent as a header:

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/1986950/167252332-3f0de869-09b3-47b4-b0d8-2125dd6dead8.png">

However, I do see a `referer` being sent as a header for the regular graphql requests:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/1986950/167252354-e6a84a39-2e71-49e6-9daa-a3a93f00ac10.png">

Is there a way to configure the client to send the referer as a header?

**Comment by enisdenjo** [comment, marked as answer]
Setting custom headers is not possible in browsers.

You can do so in Node though - for that you can consult the ["Client usage in Node with custom headers" recipe](https://github.com/enisdenjo/graphql-ws#node-client-headers).

**Reply by paymog** [reply]
Got it, thanks for the info and super quick response!
